### PR TITLE
feat: occurrence support

### LINKED ISSUE #43: Occurence support

### PR DESCRIPTION
**Pull request by SamAg19**
fixes #43

### CODE REVIEW

File: contracts/Core/BondManager.sol [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    function addJobsToCollection(
+        uint32 bondId,
+        Structs.Job[] memory jobs,
+        int8 collectionPower,
+        uint32 collectionTolerance,
+        uint32 collectionAggregation
+    ) external databondCreatorCheck(bondId, msg.sender) {
+        uint32 epoch = getEpoch();
```
Comment (SamAg19) [inline contracts/Core/BondManager.sol:117]: minJobs as well

Comment (SamAg19) [inline contracts/Core/BondManager.sol:117]: minJobs as well

File: contracts/Core/CollectionManager.sol [hunk 2]
```diff
@@ -244,6 +233,52 @@ contract CollectionManager is Initializable, CollectionStorage, StateManager, Co
...
+        uint32[] memory medians
+    ) external override onlyRole(COLLECTION_CONFIRMER_ROLE) {
+        bool toBeUpdated = false;
+        for (uint256 i = 0; i < blockIds.length; i++) {
+            uint16 collectionId = blockIds[i];
+            collections[collectionId].result = medians[i];
+            collections[collectionId].epochLastReported = epoch;
+            if (collections[collectionId].epochLastReported + collections[collectionId].occurrence != epoch + 1) {
+                // slither-disable-next-line costly-loop
+                numActiveCollections = numActiveCollections - 1;
```
Comment (SamAg19) [inline contracts/Core/CollectionManager.sol:248]: follow good-bad implementation provided by slither

Comment (SamAg19) [inline contracts/Core/CollectionManager.sol:248]: follow good-bad implementation provided by slither

File: contracts/Core/BondManager.sol [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        }
+        databonds[bondId].active = databondStatus;
+        databonds[bondId].epochBondLastUpdated = getEpoch();
+        collectionManager.setCollectionStatus(databondStatus, databonds[bondId].collectionId);
+    }
+
+    function setOccurrence() external override onlyRole(OCCURRENCE_CALCULATOR_ROLE) {
+        for (uint32 i = 1; i <= numDataBond; i++) {
+            if (databonds[i].active && databonds[i].bond >= minBond) {
+                uint256 occurrence = (databonds[i].jobIds.length * depositPerJob) / databonds[i].bond;
```
Comment (SamAg19) [inline contracts/Core/BondManager.sol:211]: customisable bond not taken into consideration anymore here. check for desired occurrence is greater than permissible occurrence

Comment (SamAg19) [inline contracts/Core/BondManager.sol:211]: customisable bond not taken into consideration anymore here. check for desired occurrence is greater than permissible occurrence

File: contracts/Core/BondManager.sol [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        // slither-disable-next-line timestamp
+        require(databonds[bondId].epochBondLastUpdated + epochLimitForUpdateBond <= epoch, "invalid databond update");
+
+        databonds[bondId].epochBondLastUpdated = epoch;
+        databonds[bondId].jobIds = jobIds;
+
+        collectionManager.updateCollection(collectionId, tolerance, aggregationMethod, power, jobIds);
+    }
+
+    function addJobsToCollection(
```
Comment (SamAg19) [inline contracts/Core/BondManager.sol:110]: calculate occurrence in places where databond collections are bing directly affected

Comment (SamAg19) [inline contracts/Core/BondManager.sol:110]: calculate occurrence in places where databond collections are bing directly affected

**Comment by SamAg19** [comment]
Will work on the changes mentioned once issue #801 is fixed.

**Comment by SkandaBhat** [comment]
Please add natspec comments on Bond Manager @SamAg19

**Comment by SkandaBhat** [comment]
@SamAg19 there have been multiple post deployment script errors reported by @shekhar2807 . Please fix them on priority.

**Pull request merged**